### ISSUE TITLE
Retaining Tab Indexing

### ISSUE THREAD

**Issue by joshuarubin** [glacambre/firenvim#1270]
I'm trying to find a way to retain functionality so that when I use `<Tab>` or `<S-Tab>` the browser moves focus as it would without firenvim. I've tried ignoring the tab key. I've also tried adding mappings that run `hide_frame()` and `focus_page()` but I don't think I can get it to work with those. I don't want to hide the frame and I don't want the whole page focused, just the next (or previous) tab index. I also tried `eval_js()` but CSP issues prevent that from working in too many places. Any advice would be appreciated, thanks!

**Comment by glacambre** [comment]
Is `nnoremap <Tab> <Cmd>call firenvim#focus_input()<CR>` good enough for your use case?

**Comment by joshuarubin** [comment]
I would really like to tab to the _next_ (or previous) item in the list. It would really make using firenvim far more transparent for me.

**Comment by glacambre** [comment]
Makes sense. I'm not sure I'll be able to 100% match the browser's tab-cycling logic (and I'm sure it differs from browser to browser...) but I can probably make something that gets close :).

**Comment by glacambre** [comment]
@joshuarubin I've implemented this feature, and I think it's pretty close to how browsers behave, but I'd still like you to check if it behaves as you expect it to.

In order to try this feature, follow these steps:
- Remove Firenvim from your browser
- If you're on chrome, download [chrome.tar.gz](https://github.com/glacambre/firenvim/files/7796764/chrome.tar.gz), extract it, go to chrome://extensions, select "load unpacked" and select the directory you extracted.
- If you're on firefox, download [firefox.tar.gz](https://github.com/glacambre/firenvim/files/7796765/firefox.tar.gz), extract it, go to about://debugging, select "this firefox" in the column on the left and then select "load temporary add-on".

Once this new Firenvim version is downloaded, go to the website you would usually visit, trigger Firenvim and once the Firenvim frame is focused, run `:call rpcnotify(firenvim#get_chan(), 'firenvim_focus_next')` and `:call rpcnotify(firenvim#get_chan(), 'firenvim_focus_prev')` to select either the next element or the previous one.

Please let me know how it goes and don't forget to remove this test add-on and to switch back to the chrome web store/addons.mozilla.org version once you've finished testing :)

**Comment by dmfay** [comment]
Not Joshua but I've been hoping for this too. I added these mappings:

```
if exists('g:started_by_firenvim')
  nnoremap <Tab> <Cmd>:call rpcnotify(firenvim#get_chan(), 'firenvim_focus_next')<CR>
  nnoremap <S-Tab> <Cmd>:call rpcnotify(firenvim#get_chan(), 'firenvim_focus_prev')<CR>
endif
```

There are still some oddities easily demonstrated on [txti.es](http://txti.es) after opening firenvim in the textarea.

Tabbing from firenvim repeatedly cycles between the terms of service link, the save and options buttons at the bottom, and back to firenvim -- never focusing the URL bar or the links above firenvim.

Shift+tab from firenvim cycles between the "how to use" link, the URL bar, and firenvim, never focusing _below_ and missing the "what is" and homepage links.

I should note that the header and "what is" links both show as visited to me while the "how to use" and "terms of service" are unvisited. Might be a coincidence, might not.

**Comment by glacambre** [comment]
@dmfay Thanks a lot for trying this feature out and for reporting the issue you found. It's now fixed in #1278 - now I only need to write a couple of tests and I'll be able to merge the feature and publish a new version to the chrome and firefox add-on stores :)

**Comment by joshuarubin** [comment]
This is great, thank you so much!
